### ISSUE TITLE
Can't write to process [MacOS]

### ISSUE THREAD

**Issue by AOx0** [johnsonjason/vmemory#1]
I'm able to read memory from a process but fail to write to it on MacOS Monterey 12.0.1. I'm using the second example template.

The program panics at `memory_darwin.rs:43:103`

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: 2', /Users/alejandro/.cargo/registry/src/github.com-1ecc6299db9ec823/vmemory-0.1.0/src/memory_darwin.rs:43:103
```

Specifically at :

```rust
proc_protect(target_task, _address, buffer.len(), VM_PROT_READ | VM_PROT_WRITE | VM_PROT_EXECUTE).unwrap();
```

```
  17:        0x10825f56e - vmemory::memory_darwin::write_memory::h357da045cd595797
                               at /Users/alejandro/.cargo/registry/src/github.com-1ecc6299db9ec823/vmemory-0.1.0/src/memory_darwin.rs:43:5
  18:        0x10825f311 - vmemory::ProcessMemory::write_memory::h653137f08c1e7ebd
                               at /Users/alejandro/.cargo/registry/src/github.com-1ecc6299db9ec823/vmemory-0.1.0/src/lib.rs:452:13
  19:        0x10825e6d5 - viewMemory::main::h779eba0e797e2b3b
                               at /Users/alejandro/viewMemory/src/main.rs:23:2
```

**Comment by johnsonjason** [comment]
I tested with Big Sur prior, but you have to be running as root and certain processes will continue to fail - even if you are root because of System Integrity Protection (for example, attaching to any code-signed app unfortunately, even proper debuggers like GDB will fail.). Assuming you took the steps to enable root / disable SIP:

It looks like this might be due to another error and possibly fixable(?): https://opensource.apple.com/source/xnu/xnu-1228/osfmk/mach/kern_return.h

```C
#define KERN_PROTECTION_FAILURE		2
		/* Specified memory is valid, but does not permit the
		 * required forms of access.
		 */
```

I imagine it has to do with the memory region's maximum protection flag and I do not have the understanding of Mach's architecture to suggest a valid fix, but it looks like the maximum protection can only be reduced, meaning, the region's protection can only be modified to be more restrictive, not less restrictive.

Maybe it is possible to suspend the task, read the region, unmap the region with [mach_vm_deallocate](https://developer.apple.com/documentation/kernel/1402285-mach_vm_deallocate) and remap it with a new protection?

**Comment by johnsonjason** [comment]
I actually just encountered this same error (2) today when running some tests. For my case, it appeared it as due to the maximum_protection on the memory region. I tested out what I suggested above about remapping the memory region and it appears to have worked. I pushed the change and can be seen in [memory_darwin.rs](https://github.com/johnsonjason/vmemory/blob/b954eaecf21be3e34debfc985188ef3999af0697/src/memory_darwin.rs#L199)

Hopefully that will work for your case too.
